File: exp2/kruskal.cc
#include <iostream>
#include <algorithm>
#include <cstring>

using namespace std;

const int maxn = 1000;

struct Edge {
    int x;
    int y;
    int weight;
} edges[maxn];

int parent[maxn];
int depth[maxn];

void makeSet() {
    memset(parent, -1, sizeof(parent));
    memset(depth, 0, sizeof(depth));
}

int find(int x) {
    if (parent[x] == -1) {
        return x;
    }
    parent[x] = find(parent[x]);
    return parent[x];
}

void Union(int x, int y) {
    int fx = find(x);
    int fy = find(y);

    if (depth[fx] > depth[fy]) {
        parent[fy] = fx;
    } else {
        parent[fx] = fy;
        if (depth[fx] == depth[fy]) {
            ++depth[fy];
        }
    }
}

int main() {
    int n, m;
    cin >> n >> m;
    for (int i = 0; i < m; ++i) {
        cin >> edges[i].x >> edges[i].y >> edges[i].weight;
    }
    sort(edges, edges + m, [](struct Edge a, struct Edge b) {
        return a.weight < b.weight;
    });
    makeSet();
    for (int i = 0; i < m; ++i) {
        int fx = find(edges[i].x);
        int fy = find(edges[i].y);
        if (fx != fy) {
            cout << edges[i].x << " -> " << edges[i].y << " : " << edges[i].weight << endl;
            Union(edges[i].x, edges[i].y);
        }
    }
}

File: exp2/kruskal.go
package main

import (
	"fmt"
	"os"
	"sort"
)

const INFILE = "/Users/coordinate36/program/algorithm/exp2/kruskal.txt"

type Edge struct {
	X, Y   int
	Weight int
}

type Edges []Edge

var edges []Edge
var depth []int
var root []int

func (e Edges) Len() int           { return len(e) }
func (e Edges) Less(i, j int) bool { return e[i].Weight < e[j].Weight }
func (e Edges) Swap(i, j int)      { e[i], e[j] = e[j], e[i] }

func find(x int) int {
	if root[x] == -1 {
		return x
	}

	root[x] = find(root[x])
	return root[x]
}

func Union(x, y int) {
	fx := find(x)
	fy := find(y)

	if depth[fx] < depth[fy] {
		root[fx] = fy
	} else {
		root[fy] = fx
		if depth[fx] == depth[fy] {
			depth[fx]++
		}
	}
}

func kruskal() {
	for _, e := range edges {
		fx := find(e.X)
		fy := find(e.Y)
		if fx != fy {
			fmt.Printf("%d -> %d: %d\n", e.X, e.Y, e.Weight)
			Union(fx, fy)
		}
	}
}

func input() {
	var m, n int
	fmt.Scanf("%d%d", &n, &m)
	edges = make([]Edge, m)
	depth = make([]int, n+1)
	root = make([]int, n+1)
	for i := range root {
		root[i] = -1
	}

	for i := 0; i < m; i++ {
		fmt.Scanf("%d%d%d", &edges[i].X, &edges[i].Y, &edges[i].Weight)
	}

	fmt.Scanln()
}

func main() {

	// redirect stdin
	inFile, err := os.Open(INFILE)
	if err != nil {
		panic(err)
	}
	defer inFile.Close()
	os.Stdin = inFile

	var numTests int
	fmt.Scanf("%d", &numTests)
	fmt.Scanln()

	for i := 0; i < numTests; i++ {
		input()
		sort.Sort(Edges(edges))
		fmt.Printf("Test%d\n", i)
		kruskal()
		fmt.Println()
	}
}
